cirq-google/cirq_google/engine/abstract_processor.py

```diff
@@ -0,0 +1,420 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+import datetime
+
+from typing import Dict, Iterable, List, Optional, Sequence, TYPE_CHECKING, Union
+
+import cirq
+from cirq_google.engine.client.quantum import types as qtypes
+from cirq_google.engine.client.quantum import enums as qenums
+from cirq_google.api import v2
+from cirq_google.engine import calibration
+
+if TYPE_CHECKING:
+    import cirq_google
+    from cirq_google.engine.abstract_engine import AbstractEngine
+    from cirq_google.engine.abstract_job import AbstractJob
+    from cirq_google.engine.abstract_program import AbstractProgram
+    from cirq_google.serialization.serializer import Serializer
+
+
+class AbstractProcessor(ABC):
+    """An abstract interface for a quantum processor.
+
+    This quantum processor has the ability to execute single circuits
+    (via the run method), parameter sweeps (via run_sweep), batched
+    lists of circuits (via run_batch), and calibration
+    requests (via run_calibration).  Running circuits can also be
+    done using the `cirq.Sampler` by calling get_sampler.
+
+    The processor interface also includes methods to create, list,
+    and remove reservations on the processor for dedicated access.
+    The processor can also list calibration metrics for the processor
+    given a time period.
+
+    This is an abstract class.  Inheritors should implement abstract methods.
+    """
+
+    def run(
+        self,
+        program: cirq.Circuit,
+        program_id: Optional[str] = None,
+        job_id: Optional[str] = None,
+        param_resolver: cirq.ParamResolver = cirq.ParamResolver({}),
```

Comment:
This seems like an unsafe default to have. Could we use `None` here instead and auto promote it to an empty paramresolver ?